server/modules/user/service/userService.js

```diff
@@ -1,15 +1,51 @@
+const aws = require('../../../system/aws')
+
 const UserManager = require('../manager/userManager')
 
+const AuthGroupManager = require('../../auth/manager/authManager')
+
+const UserValidator = require('../../../../common/user/userValidator')
+const User = require('../../../../common/user/user')
+const AuthGroups = require('../../../../common/auth/authGroups')
+
+const SystemError = require('../../../utils/systemError')
+
+const inviteUser = async (user, surveyId, email, groupId) => {
+  const dbUser = await UserManager.fetchUserByEmail(email)
+
+  if (dbUser) {
+    const newUserId = User.getId(dbUser)
+    const newUserGroups = User.getAuthGroups(dbUser)
+    const hasRoleInSurvey = newUserGroups.some(g => AuthGroups.getSurveyId(g) === surveyId)
+
+    if (hasRoleInSurvey) {
+      throw new SystemError('userHasRole')
+    } else {
+      await AuthGroupManager.insertUserGroup(user, surveyId, groupId, newUserId)
+      return dbUser
+    }
+  } else {
+    const newUser = await UserManager.insertUser(user, surveyId, email, groupId)
+    await aws.inviteUser(email)
+
```

Comment:
why don't we first wait for the successful response from AWS cognito and then insert user into the db?

Reply:
Yes, it has already been done on the accept-invitation branch